server/src/main/java/org/apache/druid/discovery/DruidNodeDiscoveryProvider.java

```diff
@@ -139,45 +150,39 @@ public void nodesAdded(List<DiscoveryDruidNode> nodesDiscovered)
             }
           }
 
-          ImmutableList<DiscoveryDruidNode> newNodesAdded = null;
-          if (uninitializedNodeTypeListeners.isEmpty()) {
-            newNodesAdded = builder.build();
-          } else if (uninitializedNodeTypeListeners.remove(this) && uninitializedNodeTypeListeners.isEmpty()) {
-            newNodesAdded = ImmutableList.copyOf(nodes.values());
-          }
-
-          if (newNodesAdded != null) {
+          if (!nodesAdded.isEmpty()) {
+            Collection<DiscoveryDruidNode> unmodifiableNodesAdded = Collections.unmodifiableCollection(nodesAdded);
             for (Listener listener : listeners) {
               try {
-                listener.nodesAdded(newNodesAdded);
+                listener.nodesAdded(unmodifiableNodesAdded);
               }
               catch (Exception ex) {
-                log.error(ex, "Listener[%s].nodesAdded(%s) threw exception. Ignored.", listener, newNodesAdded);
+                log.error(ex, "Listener[%s].nodesAdded(%s) threw exception. Ignored.", listener, nodesAdded);
               }
             }
           }
         }
       }
 
       @Override
-      public void nodesRemoved(List<DiscoveryDruidNode> nodesDisappeared)
+      public void nodesRemoved(Collection<DiscoveryDruidNode> nodesDisappeared)
       {
         synchronized (lock) {
-          ImmutableList.Builder<DiscoveryDruidNode> builder = ImmutableList.builder();
+          List<DiscoveryDruidNode> nodesRemoved = new ArrayList<>();
           for (DiscoveryDruidNode node : nodesDisappeared) {
             DiscoveryDruidNode prev = nodes.remove(node.getDruidNode().getHostAndPortToUse());
             if (prev != null) {
-              builder.add(node);
+              nodesRemoved.add(node);
             } else {
               log.warn("Node[%s] disappeared but was unknown for service listener [%s].", node, service);
             }
           }
 
-          if (uninitializedNodeTypeListeners.isEmpty()) {
-            ImmutableList<DiscoveryDruidNode> nodesRemoved = builder.build();
+          if (!nodesRemoved.isEmpty()) {
```

Comment:
Before that we notified on removed nodes after all types of nodes were initialized, now we just ignore?

Reply:
No, removes are always reported now (even before full initialization if they happen), it's just that empty update lists are skipped. Refactored code a bit so that it should be clearer.

Reply:
So not reporting removed nodes before initialization is a mistake?

Reply:
It was not a mistake. And now there is no mistake, because `nodeViewInitialized()` is not tied with `nodesAdded()`.

Reply:
We ignored `removedNodes` until all node type listeners have received first `nodesAdded` event (old version), (new version) we always propagate (non-empty) `removedNodes` even if we haven't received `initialized` for some node types, right?

Reply:
Right

Reply:
So it sounds like it was a mistake before, as we didn't propagate some events for node type A while some other node B wasn't initialized (effectively losing update for node type A), right?

Reply:
No, there was no mistake. nodesRemoved were not reported to listeners before full initialization, but the changes were still accumulated in `nodes`.